### ISSUE TITLE
WriteCommandError diagnostic event missing when using ExecuteReader when an OUTPUT clause is present

### ISSUE THREAD

**Issue by mattjohnsonpint** [dotnet/SqlClient#1839]
When an `OUTPUT` clause is present in a query, such as in the following:
```sql
INSERT INTO [TestEntities] ([Property]) OUTPUT INSERTED.[Id] VALUES ('TestValue')
```
If that query is run using `ExecuteReader` (or `ExecuteReaderAsync`), any error created (such as violating a constraint) is not emitted as a diagnostic source event.  Normally, the `Microsoft.Data.SqlClient.WriteCommandError` event would be emitted in such a case.

The error *is* emitted if the `OUTPUT` clause is not in the query, or if `ExecuteScalar` is used instead.

This is a problem because EF Core 7 has changed behavior to use an `OUTPUT` clause where it used to do a `SELECT` with `scope_identity()`, and it uses `ExecuteReader` because it sometimes asks for multiple return values.  I originally brought this up as an EF Core 7 bug, but I believe it's actually a problem with SQL Client.  See https://github.com/dotnet/efcore/issues/29543 for more detials.

### To reproduce

- Clone the repository at: https://github.com/mattjohnsonpint/sqlclient-diagnosticlistner-issue-repro
- Run the application and examine the output.

The application writes three rows to a table that has an identity column and another column that has a unique index.  Because it uses the same value for each write, only the first insert succeeds.  The second and third inserts fail due to the unique index being violated.  When doing so, only the second insert is correctly logged by the diagnostic listener included with the app.  The third insert is not logged.  The only difference between them is that one omits the `OUTPUT` clause.

The app emits:
```
Inserting row...
Row inserted with ID = 1.
Inserting row...
Error: Cannot insert duplicate key row in object 'dbo.TestEntities' with unique index 'IX_TestEntities_Property'. The duplicate key value is (TestValue).
The statement has been terminated.
Command: INSERT INTO [TestEntities] ([Property]) VALUES ('TestValue')
Inserting row...
```

It *should* emit:
```
Inserting row...
Row inserted with ID = 1.
Inserting row...
Error: Cannot insert duplicate key row in object 'dbo.TestEntities' with unique index 'IX_TestEntities_Property'. The duplicate key value is (TestValue).
The statement has been terminated.
Command: INSERT INTO [TestEntities] ([Property]) VALUES ('TestValue')
Inserting row...
Error: Cannot insert duplicate key row in object 'dbo.TestEntities' with unique index 'IX_TestEntities_Property'. The duplicate key value is (TestValue).
The statement has been terminated.
Command: INSERT INTO [TestEntities] ([Property]) VALUES ('TestValue')
```

The source code of the demo app is well commented, and also shows that if you change from `ExecuteReader` to `ExecuteScalar`, then output is as expected.

**Comment by roji** [comment]
To repeat a point I made in https://github.com/dotnet/efcore/issues/29543, expecting ExecuteReader to always emit error diagnostics when a command fails already doesn't work for batched commands; with `INSERT ...; INSERT ...`, ExecuteReader doesn't emit error diagnostics if the 2nd INSERT fails. So there already isn't a guarantee around this.

**Comment by mattjohnsonpint** [comment]
Perhaps, but the case I showed is not using batched commands. It's just a single command each time.

Or are you saying that diagnostic source error events aren't guaranteed to be present at all?

**Comment by roji** [comment]
> Or are you saying that diagnostic source error events aren't guaranteed to be present at all?

Yeah. I believe that if diagnostic source errors aren't emitted in the batched case, their value is somewhat limited also in the non-batched case (either a diagnostic always works or, well, you can't depend on it). But it may make sense to look into reliable diagnostics for the batching case too.

**Comment by DavoudEshtehari** [comment]
Currently, diagnostic events are supported on `SqlConnection`, `SqlCommand`, and `SqlTrnsaction`. When an `OUTPUT` clause is used in an insert expression by calling the `ExecuteReader()` function, the exception defers to the `SqlDataReader.Read()` function call, which is out of scope of diagnostic support in SqlClient at the moment.
Thank you @mattjohnsonpint for bringing it up. I believe it can be a good idea for the future.

**Comment by jamescrosswell** [comment]
@DavoudEshtehari just following up on this. Is there any chance that `SqlDataReader` will see diagnostic support in the future?

I'm looking into this issue, which is currently blocked by some solution for this 😢:
- https://github.com/getsentry/sentry-dotnet/issues/2044#issue-1447226249

**Comment by DavoudEshtehari** [comment]
@jamescrosswell I'm afraid we don't have any immediate plans to address this feature. I've marked this up for external contributors, though.
